### PR TITLE
Add a random worker selection once the pool is full strategy.

### PR DESCRIPTION
**Pull request by jerome-benoit**
Not sure it will be useful for a real world workload but it was quick to
implement ;)

Signed-off-by: Jérôme Benoit <jerome.benoit@sap.com>

## PR Checklist

- [x] Please add a description of your changes.
- [ ] We need your changes to come with unit tests in order to keep this project in quality and easy to maintain.
- [ ] Please add a link to the open issue or task that this pull request will resolve.
- [ ] Add a note to the changelog to indicate the change.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

If yes please indicate it and migration info into the CHANGELOG.md file.

---

### CODE REVIEW

**Comment by pioardi** [comment]
Thanks for your submission Jerome, I Will look more deeply tomorrow but please note that there are two security hotspots on sonar

File: src/pools/selection-strategies.ts [hunk 1]
```diff
@@ -106,25 +110,41 @@ class LessRecentlyUsedWorkerChoiceStrategy<
...
+        this.pool
       )
+      if (freeWorker) {
+        return freeWorker
+      }
+    }
+    return this.pool.workers[this.getRandomInt(0, this.pool.workers.length - 1)]
+  }
+
+  private getRandomInt (max: number, min = 0): number {
```
Comment (Shinigami92) [inline src/pools/selection-strategies.ts:143]: Should we use this implementation?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random#getting_a_random_integer_between_two_values

File: src/pools/selection-strategies.ts [hunk 2]
```diff
@@ -94,7 +98,7 @@ class LessRecentlyUsedWorkerChoiceStrategy<
     // A worker is always found because it picks the one with fewer tasks
     let lessRecentlyUsedWorker!: Worker
     for (const [worker, numberOfTasks] of this.pool.tasks) {
-      if (numberOfTasks === 0) {
+      if (!this.pool.isDynamic() && numberOfTasks === 0) {
```
Comment (Shinigami92) [inline src/pools/selection-strategies.ts:101]: We may extract this `!this.pool.isDynamic()` out of the loop to a `const` and safe performance this way, cause the function is not called for each task

**Review by Shinigami92 — CHANGES_REQUESTED** [review]

**Comment by pioardi** [comment]
I am not sure we really Need this guys, i am thinking to a real scenario where this can be really a benefit mmmmm

**Comment by Shinigami92** [comment]
> Thanks for your submission Jerome, I Will look more deeply tomorrow but please note that there are two security hotspots on sonar

It's because we call `Math.random()`. But we are not affected by any vulnerabilities and have not security code here. So you @pioardi can go to SonarCloud and just disable this warning.

**Comment by Shinigami92** [comment]
> I am not sure we really Need this guys, i am thinking to a real scenario where this can be really a benefit mmmmm

I also thought that :thinking:  
The code itself is okay, but the need... :shrug: Not really :thinking:

File: src/pools/selection-strategies.ts (hunk 1 above)
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:143]: The implementation in the PR is more optimized: one floor, one random only.

**Comment by jerome-benoit** [comment]
> I am not sure we really Need this guys, i am thinking to a real scenario where this can be really a benefit mmmmm

It's useful for testing purpose, especially for internal benchmarking on selection strategies as a kind of worst case.

File: src/pools/selection-strategies.ts (hunk 1 above)
Comment (Shinigami92) [reply src/pools/selection-strategies.ts]: But the need of an if else branch :thinking:  
An internal benchmark could be nice here again

Comment (Shinigami92) [reply src/pools/selection-strategies.ts]: Ah you removed the branch

**Comment by pioardi** [comment]
> > I am not sure we really Need this guys, i am thinking to a real scenario where this can be really a benefit mmmmm
> 
> It's useful for testing purpose, especially for internal benchmarking on selection strategies as a kind of worst case.

mmmmm I really don't see this needed for testing purposes, I mean why we need a worst case ?
I am sorry I just really do not see any benefit providing this feature, rather it could be misleading for end users

**Comment by Shinigami92** [comment]
I also agree with @pioardi 
Also, and now you see hopefully the problem, you are doing multiple things in one PR again.
Some little code snippets are okay but they are in this big branch and you need to extract them now into a new PR if we want them :slightly_frowning_face:  

@jerome-benoit If you provide a new feature (like a new strategy) you should also think about creating an issue for that so we can talk about it before wasting time implementing it
Instead we could implement a way that users can pass their own strategy. Currently the switch-case only allows poolifier internal strategies and prevent the user from trying they own strategy

**Comment by jerome-benoit** [comment]
> > > I am not sure we really Need this guys, i am thinking to a real scenario where this can be really a benefit mmmmm
> > 
> > 
> > It's useful for testing purpose, especially for internal benchmarking on selection strategies as a kind of worst case.
> 
> mmmmm I really don't see this needed for testing purposes, I mean why we need a worst case ?
> I am sorry I just really do not see any benefit providing this feature, rather it could be misleading for end users

It's a performance evaluation thing, it's like fault injection for security: you do things unexpected in a code path and you fix its impact. And you keep it as a landmark for non regression on the code path.   

It's ok to change the API to allow to pass that strategy for performance evaluation instead of providing it in poolifier.

**Comment by jerome-benoit** [comment]
> @jerome-benoit If you provide a new feature (like a new strategy) you should also think about creating an issue for that so we can talk about it before wasting time implementing it
> Instead we could implement a way that users can pass their own strategy. Currently the switch-case only allows poolifier internal strategies and prevent the user from trying they own strategy

It's way more efficient to discuss around a piece of code than around an issue. 
GitHub implied code flow is not efficient but I can't change it. I will try to use stgit in a future to split implementation in small logical pieces that could be submitted as a GitHub PR (the inefficient GitHub understanding of it ... unfortunately)

**Comment by Shinigami92** [comment]
@jerome-benoit I think you are confusing `efficient` with `fast`

**Comment by pioardi** [comment]
> > > > I am not sure we really Need this guys, i am thinking to a real scenario where this can be really a benefit mmmmm
> > > 
> > > 
> > > It's useful for testing purpose, especially for internal benchmarking on selection strategies as a kind of worst case.
> > 
> > 
> > mmmmm I really don't see this needed for testing purposes, I mean why we need a worst case ?
> > I am sorry I just really do not see any benefit providing this feature, rather it could be misleading for end users
> 
> It's a performance evaluation thing, it's like fault injection for security: you do things unexpected in a code path and you fix its impact. And you keep it as a landmark for non regression on the code path.
> 
> It's ok to change the API to allow to pass that strategy for performance evaluation instead of providing it in poolifier.

I think having the possibility to pass a strategy and then pass the random strategy from a test would be a great trade-off.

**Comment by jerome-benoit** [comment]
> @jerome-benoit I think you are confusing `efficient` with `fast`

Not at all, have a look at the way PRs are done on git, Linux Kernel, OpenWrt (when GitHub is not used), glibc, etc. PRs here huge and review is done at the commit level.

**Comment by Shinigami92** [comment]
> > @jerome-benoit I think you are confusing `efficient` with `fast`
> 
> Not at all, have a look at the way PRs are done on git, Linux Kernel, OpenWrt (when GitHub is not used), glibc, etc. PRs here huge and review is done at the commit level.

You compare big-sizes repos with great amount of developers with a 3-member small-sized JS-lib :eyes:

**Comment by jerome-benoit** [comment]
> > > @jerome-benoit I think you are confusing `efficient` with `fast`
> > 
> > 
> > Not at all, have a look at the way PRs are done on git, Linux Kernel, OpenWrt (when GitHub is not used), glibc, etc. PRs here huge and review is done at the commit level.
> 
> You compare big-sizes repos with great amount of developers with a 3-member small-sized JS-lib 👀

A code flow does not change because of the community size or code size, it changes because of the tooling and project rules regarding it. 
When you use GitHub, you are stuck to PR based level review and that's not efficient.

**Comment by pioardi** [comment]
@jerome-benoit  if you can jump into the chat I would love a discussion about the last message you sent :)

**Comment by Shinigami92** [comment]
IMO this PR is superseded by #234  
I will close it

**Pull request closed without merging**